Engine/DataFeeds/ApiDataProvider.cs

```diff
@@ -40,38 +60,114 @@ public class ApiDataProvider : IDataProvider
+                    $" Setting purchase limit to balance : {_balance}");
+                _purchaseLimit = _balance;
+            }
         }
 
         /// <summary>
         /// Retrieves data to be used in an algorithm.
         /// If file does not exist, an attempt is made to download them from the api
         /// </summary>
-        /// <param name="key">A string representing where the data is stored</param>
+        /// <param name="filePath">File path representing where the data requested</param>
         /// <returns>A <see cref="Stream"/> of the data requested</returns>
-        public Stream Fetch(string key)
+        public Stream Fetch(string filePath)
         {
-            Symbol symbol;
-            DateTime date;
-            Resolution resolution;
-
-            // Fetch the details of this data request
-            if (LeanData.TryParsePath(key, out symbol, out date, out resolution))
+            // If we don't already have this file or its out of date, download it
+            if (NeedToDownload(filePath))
             {
-                if (!File.Exists(key) || IsOutOfDate(resolution, key))
+                // TODO Factor files and map files do not have a price, so they come back as -1. This works for now but do we want to try and handle that case uniquely?
+                // Verify we have enough credit to handle this
+                var price = _dataPrices.GetPrice(filePath);
+                if (price > 0)
                 {
-                    return DownloadData(key, symbol, date, resolution);
+                    if (_purchaseLimit < price)
+                    {
+                        throw new Exception($"Cost {price} for {filePath} data exceeds remaining purchase limit: {_purchaseLimit}");
+                    }
+
+                    // Update our purchase limit and balance.
+                    _purchaseLimit -= price;
+                    _balance -= price;
                 }
 
-                // Use the file already on the disk
-                return new FileStream(key, FileMode.Open, FileAccess.Read, FileShare.Read);
+                return DownloadData(filePath);
+            }
+
+            // Use the file already on the disk
+            if (File.Exists(filePath))
+            {
+                return new FileStream(filePath, FileMode.Open, FileAccess.Read, FileShare.Read);
             }
 
-            Log.Error("ApiDataProvider.Fetch(): failed to parse key {0}", key);
+            Log.Error($"ApiDataProvider.Fetch(): failed to fetch data at {filePath}");
             return null;
         }
 
+        private bool NeedToDownload(string filePath)
+        {
+            // Todo: TryParsePath logs error, should we do something to mute this?
+            // For files that TryParsePath can extract symbol, date, resolution (not factor files, map files, etc)
+            if (LeanData.TryParsePath(filePath, out Symbol symbol, out DateTime date, out Resolution resolution))
```

Comment:
`LeanData.TryParsePath` is able to parse map/factor file path?

Reply:
No it can't thats why I have the else clause that just checks that file date. If you pass it a factorfile path it logs an error like so:

```
Error: System.ArgumentException: Requested value 'factor_files' was not found.
   at System.Enum.TryParseByName(RuntimeType enumType, String originalValueString, ReadOnlySpan`1 value, Boolean ignoreCase, Boolean throwOnFailure, UInt64& result)
   at System.Enum.TryParseInt32Enum(RuntimeType enumType, String originalValueString, ReadOnlySpan`1 value, Int32 minInclusive, Int32 maxInclusive, Boolean ignoreCase, Boolean throwOnFailure, TypeCode type, Int32& result)
   at System.Enum.TryParse(Type enumType, String value, Boolean ignoreCase, Boolean throwOnFailure, Object& result)
   at System.Enum.Parse(Type enumType, String value, Boolean ignoreCase)
   at QuantConnect.Util.LeanData.TryParsePath(String fileName, Symbol& symbol, DateTime& date, Resolution& resolution) in C:\Users\Colton\Desktop\Working\QuantConnect\Repos\Lean\Common\Util\LeanData.cs:line 1020
   ```

It doesn't crash but I don't love that it posts the error. Wondering if it would be best to add a optional param bool to flag posting the error. True by default but can set to false in this case.

Reply:
✅